File: core/src/main/java/com/graphhopper/reader/dem/AbstractSRTMElevationProvider.java
/*
 *  Licensed to GraphHopper GmbH under one or more contributor
 *  license agreements. See the NOTICE file distributed with this work for
 *  additional information regarding copyright ownership.
 *
 *  GraphHopper GmbH licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except in
 *  compliance with the License. You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.graphhopper.reader.dem;

import com.graphhopper.coll.GHIntObjectHashMap;
import com.graphhopper.storage.DataAccess;
import com.graphhopper.util.BitUtil;
import com.graphhopper.util.Downloader;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.net.SocketTimeoutException;

/**
 * Common functionality used when working with SRTM hgt data.
 *
 * @author Robin Boldt
 */
public abstract class AbstractSRTMElevationProvider extends TileBasedElevationProvider {
    private final int DEFAULT_WIDTH;
    private final int MIN_LAT;
    private final int MAX_LAT;
    private final int WIDTH_BYTE_INDEX = 0;
    private final int DEGREE = 1;
    // use a map as an array is not quite useful if we want to hold only parts of the world
    private final GHIntObjectHashMap<HeightTile> cacheData = new GHIntObjectHashMap<>();
    private final double precision = 1e7;
    private final double invPrecision = 1 / precision;

    public AbstractSRTMElevationProvider(String baseUrl, String cacheDir, String downloaderName, int minLat, int maxLat, int defaultWidth) {
        super(cacheDir);
        this.baseUrl = baseUrl;
        downloader = new Downloader(downloaderName).setTimeout(10000);
        this.DEFAULT_WIDTH = defaultWidth;
        this.MIN_LAT = minLat;
        this.MAX_LAT = maxLat;
    }

    // use int key instead of string for lower memory usage
    int calcIntKey(double lat, double lon) {
        // we could use LinearKeyAlgo but this is simpler as we only need integer precision:
        return (down(lat) + 90) * 1000 + down(lon) + 180;
    }

    @Override
    public void release() {
        cacheData.clear();
        if (dir != null) {
            // for memory mapped type we remove temporary files
            if (autoRemoveTemporary)
                dir.clear();
            else
                dir.close();
        }
    }

    int down(double val) {
        int intVal = (int) val;
        if (val >= 0 || intVal - val < invPrecision)
            return intVal;
        return intVal - 1;
    }

    @Override
    public double getEle(double lat, double lon) {
        // Return fast, if there is no data available
        // See https://www2.jpl.nasa.gov/srtm/faq.html
        if (lat >= MAX_LAT || lat <= MIN_LAT)
            return 0;

        lat = (int) (lat * precision) / precision;
        lon = (int) (lon * precision) / precision;
        int intKey = calcIntKey(lat, lon);
        HeightTile demProvider = cacheData.get(intKey);
        if (demProvider == null) {
            if (!cacheDir.exists())
                cacheDir.mkdirs();

            int minLat = down(lat);
            int minLon = down(lon);

            String fileName = getFileName(lat, lon);
            if (fileName == null)
                return 0;

            DataAccess heights = getDirectory().create("dem" + intKey);
            boolean loadExisting = false;
            try {
                loadExisting = heights.loadExisting();
            } catch (Exception ex) {
                logger.warn("cannot load dem" + intKey + ", error:" + ex.getMessage());
            }

            if (!loadExisting) {
                try {
                    updateHeightsFromFile(lat, lon, heights);
                } catch (FileNotFoundException ex) {
                    demProvider = new HeightTile(minLat, minLon, DEFAULT_WIDTH, DEFAULT_WIDTH, precision, DEGREE, DEGREE);
                    cacheData.put(intKey, demProvider);
                    demProvider.setHeights(heights);
                    demProvider.setSeaLevel(true);
                    // use small size on disc and in-memory
                    heights.create(10)
                            .flush();
                    return 0;
                }
            }

            int width = (int) (Math.sqrt(heights.getHeader(WIDTH_BYTE_INDEX)) + 0.5);
            if (width == 0)
                width = DEFAULT_WIDTH;

            demProvider = new HeightTile(minLat, minLon, width, width, precision, DEGREE, DEGREE);
            cacheData.put(intKey, demProvider);
            demProvider.setInterpolate(interpolate);
            demProvider.setHeights(heights);
        }

        if (demProvider.isSeaLevel())
            return 0;

        return demProvider.getHeight(lat, lon);
    }

    private void updateHeightsFromFile(double lat, double lon, DataAccess heights) throws FileNotFoundException {
        try {
            byte[] bytes = getByteArrayFromFile(lat, lon);
            heights.create(bytes.length);
            for (int bytePos = 0; bytePos < bytes.length; bytePos += 2) {
                // we need big endianess to read the SRTM files
                short val = BitUtil.BIG.toShort(bytes, bytePos);
                if (val < -1000 || val > 12000)
                    val = Short.MIN_VALUE;

                heights.setShort(bytePos, val);
            }
            heights.setHeader(WIDTH_BYTE_INDEX, bytes.length / 2);
            heights.flush();

        } catch (FileNotFoundException ex) {
            logger.warn("File not found for the coordinates for " + lat + "," + lon);
            throw ex;
        } catch (Exception ex) {
            throw new RuntimeException("There was an issue looking up the coordinates for " + lat + "," + lon, ex);
        }
    }

    private byte[] getByteArrayFromFile(double lat, double lon) throws InterruptedException, IOException {
        String zippedURL = baseUrl + getDownloadURL(lat, lon);
        File file = new File(cacheDir, new File(zippedURL).getName());
        InputStream is;
        // get zip file if not already in cacheDir
        if (!file.exists())
            for (int i = 0; i < 3; i++) {
                try {
                    downloader.downloadFile(zippedURL, file.getAbsolutePath());
                    break;
                } catch (SocketTimeoutException ex) {
                    // just try again after a little nap
                    Thread.sleep(2000);
                } catch (FileNotFoundException ex) {
                    if (zippedURL.contains(".hgt.zip")) {
                        zippedURL = zippedURL.replace(".hgt.zip", "hgt.zip");
                    } else {
                        throw ex;
                    }
                }
            }

        return readFile(file);
    }

    protected String getPaddedLonString(int lonInt) {
        lonInt = Math.abs(lonInt);
        String lonString = lonInt < 100 ? "0" : "";
        if (lonInt < 10)
            lonString += "0";
        lonString += lonInt;
        return lonString;
    }

    protected String getPaddedLatString(int latInt) {
        latInt = Math.abs(latInt);
        String latString = latInt < 10 ? "0" : "";
        latString += latInt;
        return latString;
    }

    abstract byte[] readFile(File file) throws IOException;

    /**
     * Return the local file name without file ending, has to be lower case, because DataAccess only supports lower case names.
     */
    abstract String getFileName(double lat, double lon);

    /**
     * Returns the complete URL to download the file
     */
    abstract String getDownloadURL(double lat, double lon);

}

File: core/src/main/kotlin/com/graphhopper/reader/dem/AbstractSRTMElevationProvider.kt
/*
 *  Licensed to GraphHopper GmbH under one or more contributor
 *  license agreements. See the NOTICE file distributed with this work for
 *  additional information regarding copyright ownership.
 *
 *  GraphHopper GmbH licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except in
 *  compliance with the License. You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.graphhopper.reader.dem

import com.graphhopper.coll.GHIntObjectHashMap
import com.graphhopper.storage.DataAccess
import com.graphhopper.util.Downloader
import com.graphhopper.util.Helper
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.net.SocketTimeoutException

/**
 * Common functionality used when working with SRTM hgt data.
 *
 * @author Robin Boldt
 */
abstract class AbstractSRTMElevationProvider(
    baseUrl: String,
    cacheDir: String,
    downloaderName: String,
    private val minLatProvider: Int,
    private val maxLatProvider: Int,
    private val defaultWidth: Int
) : TileBasedElevationProvider(cacheDir) {

    private val widthByteIndex = 0
    private val degree = 1

    // use a map as an array is not quite useful if we want to hold only parts of the world
    private val cacheData = GHIntObjectHashMap<HeightTile>()
    private val precision = 1e7
    private val invPrecision = 1 / precision

    init {
        this.baseUrl = baseUrl
        downloader = Downloader().setTimeout(10000)
    }

    // use int key instead of string for lower memory usage
    internal fun calcIntKey(lat: Double, lon: Double): Int {
        // we could use LinearKeyAlgo but this is simpler as we only need integer precision:
        return (down(lat) + 90) * 1000 + down(lon) + 180
    }

    override fun release() {
        cacheData.clear()
        val dir = this.dir
        if (dir != null) {
            // for memory mapped type we remove temporary files
            if (autoRemoveTemporary)
                dir.clear()
            else
                dir.close()
        }
    }

    internal fun down(value: Double): Int {
        val intVal = value.toInt()
        if (value >= 0 || intVal - value < invPrecision)
            return intVal
        return intVal - 1
    }

    override fun getEle(lat: Double, lon: Double): Double {
        // Return fast, if there is no data available
        // See https://www2.jpl.nasa.gov/srtm/faq.html
        if (lat >= maxLatProvider || lat <= minLatProvider)
            return 0.0

        @Suppress("NAME_SHADOWING") val lat = (lat * precision).toInt() / precision
        @Suppress("NAME_SHADOWING") val lon = (lon * precision).toInt() / precision
        val intKey = calcIntKey(lat, lon)
        var demProvider: HeightTile? = cacheData.get(intKey)
        if (demProvider == null) {
            if (!cacheDir!!.exists())
                cacheDir!!.mkdirs()

            val minLat = down(lat)
            val minLon = down(lon)

            val fileName = getFileName(lat, lon)
            if (fileName == null || (Helper.isEmpty(baseUrl) && !File(fileName).exists()))
                return 0.0

            val heights = getDirectory().create("dem$intKey")
            var loadExisting = false
            try {
                loadExisting = heights.loadExisting()
            } catch (ex: Exception) {
                logger.warn("cannot load dem" + intKey + ", error:" + ex.message)
            }

            if (!loadExisting) {
                try {
                    updateHeightsFromFile(lat, lon, heights)
                } catch (ex: FileNotFoundException) {
                    demProvider = HeightTile(minLat, minLon, defaultWidth, defaultWidth, precision, degree, degree)
                    cacheData.put(intKey, demProvider)
                    demProvider.setHeights(heights)
                    demProvider.setSeaLevel(true)
                    // use small size on disc and in-memory
                    heights.create(10)
                            .flush()
                    return 0.0
                }
            }

            var width = (Math.sqrt(heights.getHeader(widthByteIndex).toDouble()) + 0.5).toInt()
            if (width == 0)
                width = defaultWidth

            demProvider = HeightTile(minLat, minLon, width, width, precision, degree, degree)
            cacheData.put(intKey, demProvider)
            demProvider.setInterpolate(interpolate)
            demProvider.setHeights(heights)
        }

        if (demProvider.isSeaLevel())
            return 0.0

        return demProvider.getHeight(lat, lon)
    }

    @Throws(FileNotFoundException::class)
    private fun updateHeightsFromFile(lat: Double, lon: Double, heights: DataAccess) {
        try {
            var zippedURL = baseUrl + getDownloadURL(lat, lon)
            val zipFile = File(cacheDir, File(zippedURL).name)
            if (!zipFile.exists()) downloadToFile(zipFile, zippedURL)
            val bytes = readFile(zipFile)
            heights.create(bytes.size.toLong())
            var bytePos = 0
            while (bytePos < bytes.size) {
                var value = toShort(bytes, bytePos)
                if (value < -1000 || value > 12000)
                    value = Short.MIN_VALUE

                heights.setShort(bytePos.toLong(), value)
                bytePos += 2
            }
            heights.setHeader(widthByteIndex, bytes.size / 2)
            heights.flush()

        } catch (ex: FileNotFoundException) {
            logger.warn("File not found " + heights + " for the coordinates " + lat + "," + lon)
            throw ex
        } catch (ex: Exception) {
            throw RuntimeException("There was an issue with " + heights + " looking up the coordinates " + lat + "," + lon, ex)
        }
    }

    // we need big endianess to read the SRTM files
    internal fun toShort(b: ByteArray, offset: Int): Short {
        return ((b[offset].toInt() and 0xFF shl 8) or (b[offset + 1].toInt() and 0xFF)).toShort()
    }

    @Throws(InterruptedException::class, IOException::class)
    private fun downloadToFile(file: File, zippedURLIn: String) {
        var zippedURL = zippedURLIn
        for (i in 0 until 3) {
            try {
                downloader!!.downloadFile(zippedURL, file.absolutePath)
                break
            } catch (ex: SocketTimeoutException) {
                // just try again after a little nap
                Thread.sleep(2000)
            } catch (ex: FileNotFoundException) {
                if (zippedURL.contains(".hgt.zip")) {
                    zippedURL = zippedURL.replace(".hgt.zip", "hgt.zip")
                } else {
                    throw ex
                }
            }
        }
    }

    protected fun getPaddedLonString(lonInt: Int): String {
        val lon = Math.abs(lonInt)
        var lonString = if (lon < 100) "0" else ""
        if (lon < 10)
            lonString += "0"
        lonString += lon
        return lonString
    }

    protected fun getPaddedLatString(latInt: Int): String {
        val lat = Math.abs(latInt)
        var latString = if (lat < 10) "0" else ""
        latString += lat
        return latString
    }

    @Throws(IOException::class)
    protected abstract fun readFile(file: File): ByteArray

    /**
     * Return the local file name without file ending, has to be lower case, because DataAccess only supports lower case names.
     */
    protected abstract fun getFileName(lat: Double, lon: Double): String?

    /**
     * Returns the complete URL to download the file
     */
    protected abstract fun getDownloadURL(lat: Double, lon: Double): String
}
